### DISCUSSION TITLE
How do conect tcpclient code to watsontcp

### DISCUSSION THREAD

**Discussion by MohMehrnia** [jchristn/WatsonTcp#295] [General]
My Code

```
tcpclnt = new TcpClient();

                    tcpclnt.Connect(txtIPAddress.Text, Convert.ToInt32(txtPortNumber.Text));

                    stm = tcpclnt.GetStream();

                    SetOpenIPButtonText("Close IP");

                    while (true)
                    {
                        if (stm.CanRead)
                        {
                            byte[] byteArray = new byte[100];
                            stm.Read(byteArray, 0, 100);
                            RecData = System.Text.Encoding.UTF8.GetString(byteArray);

                            Console.WriteLine("TCP: " + RecData);

                            if (RecData.IndexOf("relays=") >=0)
                            {
                                int start = RecData.IndexOf("relays=") + 7;
                                int len = RecData.IndexOf("\n", start)-start;
                                
                                int relays = Convert.ToInt32(RecData.Substring(start, len), 16);

                                SetRelays(relays);
                            }
                        }
                        Thread.Sleep(1);
                    }
```

how do convert to watson tcp?

**Comment by jchristn** [comment]
Hi @MohMehrnia there's a big difference between building your own TCP implementation (what you're doing) and using an implementation that manages TCP for you (like Watson).  WatsonTcp will fire an event (see `Events.MessageReceived`) when a message is received in its entirety, with the complete message, whereas an implementation managing TCP directly will have to continually read from the stream (e.g. `stm.Read(byteArray, 0, 100);` in your code above) and identify message demarcation points.  

I'd recommend cloning the repo and checking out the `Test.Server` and `Test.Client` package for familiarity on how the library works.  There's not a simple way to convert existing code from a TCP implementation to a managed implementation like WatsonTcp.

**Reply by MohMehrnia** [reply]
You can suggest another nuget package simplest tcp client
